src/server/managers/AppExternalComponentManager.ts

```diff
@@ -1,79 +1,119 @@
+     */
+    public getAppTouchedExternalComponents(): Map<string, Map<string, IExternalComponent>> {
+        return this.appTouchedExternalComponents;
+    }
+    /**
+     * Get all external components of an app by specifying the appId.
      *
      * @param appId the id of the app
-     * @param externalComponent the external component to register
-     * @param addToProvidedComponents whether also add the external
-     * component to the providedComponents, the value is false by default.
      */
-    public registerComponent(appId: string, externalComponent: IExternalComponent, addToProvidedComponents: boolean = false): void {
-        this.appTouchedComponents.set(appId, externalComponent);
-
-        if (addToProvidedComponents) {
-            this.providedComponents.set(appId, externalComponent);
+    public getExternalComponents(appId: string): Map<string, IExternalComponent> {
+        if (this.appTouchedExternalComponents.has(appId)) {
+            return this.appTouchedExternalComponents.get(appId);
         }
+
+        return null;
     }
     /**
-     * Remove the external component from the providedComponents by appId.
+     * Add an external component to the appTouchedExternalComponents.
+     * If you call this method twice and the component
+     * has the same name as before, the first one will be
+     * overwritten as the names provided **must** be unique.
      *
      * @param appId the id of the app
+     * @param externalComponent the external component need to be added
      */
-    public unregisterComponent(appId: string): void {
-        if (this.providedComponents.has(appId)) {
-            this.providedComponents.delete(appId);
+    public addExternalComponent(appId: string, externalComponent: IExternalComponent): void {
+        if (externalComponent.appId !== appId) {
+            throw new ExternalComponentNotMatchWithAppError();
+        }
+
+        if (!this.appTouchedExternalComponents.get(appId)) {
+            this.appTouchedExternalComponents.set(appId, new Map(Object.entries({ [externalComponent.name]: externalComponent})));
+        } else {
+            const appExternalComponents = this.appTouchedExternalComponents.get(appId);
+
+            appExternalComponents.set(externalComponent.name, externalComponent);
         }
     }
     /**
-     * Get the external component by the appId.
+     * Add enabled apps' external components from the appTouchedExternalComponents
+     * to the registeredExternalComponents.
      *
      * @param appId the id of the app
      */
-    public getExternalComponent(appId: string): IExternalComponent {
-        if (this.appTouchedComponents.has(appId)) {
-            return this.appTouchedComponents.get(appId);
+    public registerExternalComponents(appId: string): void {
+        if (!this.appTouchedExternalComponents.has(appId)) {
+            return;
         }
+        const externalComponents = this.appTouchedExternalComponents.get(appId);
```

Comment:
What happens if this map is empty? Maybe return out early if the touched external components map is empty for some reason.

Reply:
Thanks! Will do it as you suggested.

Reply:
Resolved in https://github.com/RocketChat/Rocket.Chat.Apps-engine/pull/150/commits/7303737cbec3e4033116a3074fddecb2d530d435.